stacks-signer/src/stacks_client.rs

```diff
@@ -98,8 +142,157 @@ impl StacksClient {
+        chain_id: u32,
+    ) -> Vec<u8> {
+        let mut sender_spending_condition = TransactionSpendingCondition::new_singlesig_p2pkh(
+            StacksPublicKey::from_private(sender),
+        )
+        .expect("Failed to create p2pkh spending condition from public key.");
+        sender_spending_condition.set_nonce(sender_nonce);
+
+        let auth = match (payer, payer_nonce) {
+            (Some(payer), Some(payer_nonce)) => {
+                let mut payer_spending_condition =
+                    TransactionSpendingCondition::new_singlesig_p2pkh(
+                        StacksPublicKey::from_private(payer),
+                    )
+                    .expect("Failed to create p2pkh spending condition from public key.");
+                payer_spending_condition.set_nonce(payer_nonce);
+                payer_spending_condition.set_tx_fee(tx_fee);
+                TransactionAuth::Sponsored(sender_spending_condition, payer_spending_condition)
+            }
+            _ => {
+                sender_spending_condition.set_tx_fee(tx_fee);
+                TransactionAuth::Standard(sender_spending_condition)
+            }
+        };
+        let mut unsigned_tx = StacksTransaction::new(version, auth, payload);
+        unsigned_tx.anchor_mode = anchor_mode;
+        unsigned_tx.post_condition_mode = TransactionPostConditionMode::Allow;
+        unsigned_tx.chain_id = chain_id;
+
+        let mut tx_signer = StacksTransactionSigner::new(&unsigned_tx);
+        tx_signer.sign_origin(sender).unwrap();
+        if let (Some(payer), Some(_)) = (payer, payer_nonce) {
+            tx_signer.sign_sponsor(payer).unwrap();
+        }
+
+        let mut buf = vec![];
+        tx_signer
+            .get_tx()
+            .unwrap()
+            .consensus_serialize(&mut buf)
+            .unwrap();
+        buf
+    }
 }
 
+impl InteractWithStacksContracts for StacksClient {
+    fn transaction_contract_call(
+        sender: &StacksPrivateKey,
+        nonce: u64,
+        contract_addr: &StacksAddress,
+        contract_name: &str,
+        function_name: &str,
+        function_args: &[ClarityValue],
+        tx_version: TransactionVersion,
+        chain_id: u32,
+    ) -> Vec<u8> {
+        let contract_name = ContractName::from(contract_name);
+        let function_name = ClarityName::from(function_name);
+
+        let payload = TransactionContractCall {
+            address: contract_addr.clone(),
+            contract_name,
+            function_name,
+            function_args: function_args.iter().map(|x| x.clone()).collect(),
+        };
+
+        let tx_fee = 0;
+
+        Self::seralize_sign_tx_anchor_mode_version(
+            payload.into(),
+            sender,
+            None,
+            nonce,
+            None,
+            tx_fee,
+            TransactionAnchorMode::OnChainOnly,
+            tx_version,
+            chain_id,
+        )
+    }
+
+    fn submit_tx(http_origin: &str, tx: &Vec<u8>) -> String {
```

Comment:
I don't know if this is ever going to be a concern, but if you're registering a BNS name, then you'd also need to supply an `Attachment` which contains the name's zonefile data.

Reply:
I am not entirely sure what this means XD Forgive me...

Reply:
Ah, so there's this storage system in the Stacks node called Atlas  It's a hold-over from the Blockstack / Stacks 1.0 days.  Basically, a smart contract can emit specially-crafted Atlas events which contain the hash of a chunk of data.  The Atlas system tracks these hash announcements, and then communicates with other Stacks nodes to see if they have a chunk of data which hashes to it.  Eventually, all nodes get a copy of this "attached" data.  We use it in BNS to tie a BNS name to a zone file, as well as other off-chain data like Nostr public keys and such.

I don't think the signer is ever going to use Atlas, so it's not a big deal if you're not going to support attachments.  However, if this module becomes more generic and gets spun off into its own crate, it might become important later to add support.